src/Control/Applicative/Free.hs

```diff
@@ -96,7 +97,80 @@ instance Comonad f => Comonad (Ap f) where
   extract (Ap x y) = extract y (extract x)
   duplicate (Pure a) = Pure (Pure a)
   duplicate (Ap x y) = Ap (duplicate x) (extend (flip Ap) y)
-  
+
+#ifdef LIFTED_FUNCTOR_CLASSES
+boringEqAp :: Eq1 f => Ap f a -> Ap f b -> Bool
+#else
+boringEqAp :: (Eq1 f, Functor f) => Ap f a -> Ap f b -> Bool
+#endif
+boringEqAp (Pure _) (Pure _) = True
+boringEqAp (Ap x1 y1) (Ap x2 y2) = boringEq x1 x2 && boringEqAp y1 y2
+boringEqAp _ _ = False
+
+#ifdef LIFTED_FUNCTOR_CLASSES
+liftEqAp :: Eq1 f => (a -> b -> Bool) -> Ap f a -> Ap f b -> Bool
+#else
+liftEqAp :: (Eq1 f, Functor f) => (a -> b -> Bool) -> Ap f a -> Ap f b -> Bool
+#endif
+liftEqAp eq (Pure a1) (Pure a2) = eq a1 a2
+liftEqAp eq (Ap x1 y1) (Ap x2 y2)
+  | emptyEq x1 x2 = boringEqAp y1 y2
+  | otherwise =
+      liftEq (\a1 a2 -> liftEqAp (\g1 g2 -> eq (g1 a1) (g2 a2)) y1 y2) x1 x2
```

Comment:
It took me a while to figure out how this code works. If I understand correctly, it is:

* Checking to see if `x1` and `x2` are equal and do not have any occurrences of the last type parameter. If this is the case, it short-circuits and checks if `y1` and `y2` are equal (ignoring occurrences of the last type parameter).
* If this is not the case, it does a more expensive equality check that requires looking inside both `x1`/`x2` and `y1`/`y2`.

Is this correct? If so, please write this down in a comment somewhere, since I had to stare at this code for a bit to figure this out.